lib/Conversion/SeqToSV/SeqToSV.cpp

```diff
@@ -456,48 +456,68 @@ void SeqToSVPass::runOnOperation() {
   if (failed(applyPartialConversion(circuit, target, std::move(patterns))))
     signalPassFailure();
 
+  bool needsMemRandomization = !memsByModule.empty();
+
+  auto loc = UnknownLoc::get(context);
+  auto b = ImplicitLocOpBuilder::atBlockBegin(loc, circuit.getBody());
+  if (needsRegRandomization || needsMemRandomization) {
+    b.create<sv::MacroDeclOp>("ENABLE_INITIAL_REG_");
+    b.create<sv::MacroDeclOp>("ENABLE_INITIAL_MEM_");
+    if (needsRegRandomization) {
+      b.create<sv::MacroDeclOp>("FIRRTL_BEFORE_INITIAL");
+      b.create<sv::MacroDeclOp>("FIRRTL_AFTER_INITIAL");
+    }
+    if (needsMemRandomization) {
+      b.create<sv::MacroDeclOp>("RANDOMIZE_MEM_INIT");
+    }
+    b.create<sv::MacroDeclOp>("RANDOMIZE_REG_INIT");
+    b.create<sv::MacroDeclOp>("RANDOMIZE");
+    b.create<sv::MacroDeclOp>("RANDOMIZE_DELAY");
+    b.create<sv::MacroDeclOp>("RANDOM");
+    b.create<sv::MacroDeclOp>("INIT_RANDOM");
+    b.create<sv::MacroDeclOp>("INIT_RANDOM_PROLOG_");
+  }
+
   bool hasRegRandomization = needsRegRandomization && !disableRegRandomization;
-  bool hasMemRandomization = !memsByModule.empty() && !disableMemRandomization;
+  bool hasMemRandomization = needsMemRandomization && !disableMemRandomization;
   if (!hasRegRandomization && !hasMemRandomization)
     return;
 
   // Build macros for FIRRTL-style register and memory initialization.
   // Insert them at the start of the module, after any other verbatims.
-  auto loc = UnknownLoc::get(context);
-  auto b = ImplicitLocOpBuilder::atBlockBegin(loc, circuit.getBody());
   for (Operation &op : *circuit.getBody()) {
-    if (!isa<sv::VerbatimOp>(&op)) {
+    if (!isa<sv::VerbatimOp, sv::IfDefOp>(&op)) {
       b.setInsertionPoint(&op);
       break;
     }
   }
 
+  // Create SYNTHESIS/VERILATOR macros if other passes have not done so already.
+  StringSet<> symbols;
+  for (auto sym : circuit.getOps<sv::MacroDeclOp>())
+    symbols.insert(sym.getName());
+  if (!symbols.count("SYNTHESIS"))
+    b.create<sv::MacroDeclOp>("SYNTHESIS");
+  if (!symbols.count("VERILATOR"))
+    b.create<sv::MacroDeclOp>("VERILATOR");
```

Comment:
Nit: consider inserting these as part of the check (`insert("..").second` -style, which hopefully StringSet supports or at least a bool return on insert) -- otherwise we leave a set around that looks like it has all active macros in it but might be missing two.

Reply:
I scoped this better